src/Storages/MergeTree/MergeTreeIndexTextPostingListCodec.cpp

```diff
@@ -0,0 +1,358 @@
+#include <Storages/MergeTree/MergeTreeIndexTextPostingListCodec.h>
+#include <Storages/MergeTree/MergeTreeIndexText.h>
+#include <Storages/MergeTree/IntegerCodecTrait.h>
+
+namespace DB
+{
+/// PostingListCodecImpl
+///
+/// A serializer/deserializer for a postings list (sorted docIDs) stored in a compact
+/// block-compressed format.
+///
+/// High-level idea:
+/// - Input docIDs are strictly increasing.
+/// - Values are encoded as deltas (gaps) and compressed in fixed-size blocks (kBlockSize).
+/// - Each compressed block is stored as: [1 byte bits-width][bitpacked payload].
+/// - A posting list may be split into "segments" (logical chunks, controlled by postings_list_block_size)
+///   to simplify metadata and to support multiple ranges per token (min/max docID per segment).
+///
+class PostingListCodecImpl
+{
+    /// Per-segment header written before each segment payload.
+    ///
+    /// - bytes: number of compressed bytes following this header (segment payload size)
+    /// - cardinality: number of postings (docIDs) in this segment
+    /// - base_value: the first docID in the segment (also used to restore deltas)
+    struct ContainerHeader
+    {
+        ContainerHeader() = default;
+        explicit ContainerHeader(size_t bytes_, uint32_t cardinality_, uint32_t base_value_)
+            : bytes(bytes_)
+            , cardinality(cardinality_)
+            , base_value(base_value_)
+        {
+        }
+
+        template<typename Out>
+        void encodeTo(Out & out) const
+        {
+            writeVarUInt(bytes, out);
+            writeVarUInt(cardinality, out);
+            writeVarUInt(base_value, out);
+        }
+
+        template<typename In>
+        void decodeFrom(In & in)
+        {
+            readOneField(bytes, in);
+            readOneField(cardinality, in);
+            readOneField(base_value, in);
+        }
+
+        template <typename F, typename In>
+        static void readOneField(F & out, In & in)
+        {
+            UInt64 v = 0;
+            readVarUInt(v, in);
+            out = static_cast<F>(v);
+        }
+
+        size_t bytes = 0;
+        uint32_t cardinality = 0;
+        uint32_t base_value = 0;
+    };
+
+    /// In-memory descriptor of one segment inside `compressed_data`.
+    ///
+    /// - count: number of postings in this segment
+    /// - compressed_data_offset: start offset in `compressed_data`
+    /// - compressed_data_size: payload size in bytes (excluding ContainerHeader)
+    /// - min/max: docID range covered by this segment.
```

Comment:
I know the term "docID" is from the information retrieval domain, but we don't really use it in ClickHouse. Let's go with "rowid".